rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/SpatialMatcher.scala

```diff
@@ -327,6 +329,38 @@ object SpatialMatcher extends SpatialMatcherInstances {
     } yield Unit
   }
 
+  private def guard(predicate: => Boolean): OptionalFreeMapWithCost[Unit] =
+    if (predicate) OptionalFreeMapWithCost.pure(()) else OptionalFreeMapWithCost.emptyMap
+
+  private def isolateState[S, F[_]: Monad](f: StateT[F, S, _]): StateT[F, S, S] =
+    for {
+      initState   <- StateT.get[F, S]
+      _           <- f
+      resultState <- StateT.get[F, S]
+      _           <- StateT.set[F, S](initState)
+    } yield resultState
+
+  def memoizeInHashMap[A, B, C](f: (A, B) => C): (A, B) => C = {
```

Comment:
This file is already too complex. We should move this function to some utility collection.

Reply:
While I agree in principle that extracting generic utils is a worthy deed, this one here has a pretty specific interface - so I'd prefer it to be treated as an implementation detail of the SpatialMatcher. It should be private though.